**Table of contents**<a id='toc0_'></a>    
- [Introducion](#toc1_)    
- [Conexion de la base de datos](#toc2_)    
- [Analisis de datos](#toc3_)    
  - [Revision de las tablas](#toc3_1_)    
    - [Tabla Books](#toc3_1_1_)    
    - [Tabla Authors](#toc3_1_2_)    
    - [Tabla publishers](#toc3_1_3_)    
    - [Tabla ratings](#toc3_1_4_)    
    - [Tabla reviews](#toc3_1_5_)    
  - [Número de libros publicados después del 1 de enero de 2000.](#toc3_2_)    
  - [Número de reseñas y calificación promedio para cada libro](#toc3_3_)    
  - [Editorial que ha publicado el mayor número de libros con más de 50 páginas](#toc3_4_)    
  - [Autor con la Calificación Promedio Más Alta de Sus Libros (Con al Menos 50 Calificaciones)](#toc3_5_)    
  - [Número Promedio de Reseñas de Texto entre Usuarios que Calificaron Más de 50 Libros](#toc3_6_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

# <a id='toc1_'></a>[Introducion](#toc0_)

El presente proyecto tiene como finalidad generar la propuesta de valor para una aplicacion dirigida a los amantes de los libros. 

# <a id='toc2_'></a>[Conexion de la base de datos](#toc0_)

In [1]:

import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# <a id='toc3_'></a>[Analisis de datos](#toc0_)

## <a id='toc3_1_'></a>[Revision de las tablas](#toc0_)

### <a id='toc3_1_1_'></a>[Tabla Books](#toc0_)

In [2]:
query= """
SELECT * FROM books LIMIT 10;
"""
books= pd.io.sql.read_sql(query, con = engine)

books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

### <a id='toc3_1_2_'></a>[Tabla Authors](#toc0_)

In [3]:
query= """
SELECT * FROM authors LIMIT 10;
"""
authors= pd.io.sql.read_sql(query, con = engine)

authors

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

### <a id='toc3_1_3_'></a>[Tabla publishers](#toc0_)

In [4]:
query= """
SELECT * FROM publishers LIMIT 10;
"""
publishers= pd.io.sql.read_sql(query, con = engine)

publishers

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

### <a id='toc3_1_4_'></a>[Tabla ratings](#toc0_)

In [5]:
query= """
SELECT * FROM ratings LIMIT 10;
"""
ratings= pd.io.sql.read_sql(query, con = engine)

ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

### <a id='toc3_1_5_'></a>[Tabla reviews](#toc0_)

In [6]:
query= """
SELECT * FROM reviews LIMIT 10;
"""
reviews= pd.io.sql.read_sql(query, con = engine)

reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## <a id='toc3_2_'></a>[Número de libros publicados después del 1 de enero de 2000.](#toc0_)

In [7]:
query = """
SELECT COUNT(book_id) AS num_books_after_2000
FROM books
WHERE publication_date > '2000-01-01';
"""
books_after_2000 = pd.read_sql(query, engine)
books_after_2000

num_books_after_2000
0                   819

Conclusiones: se han publicado 819 libros desde el 1 de enero de 2000. 

## <a id='toc3_3_'></a>[Número de reseñas y calificación promedio para cada libro](#toc0_)

In [8]:
query = """
SELECT 
    b.book_id,
    CASE WHEN review.n_reviews IS NULL THEN 0 ELSE review.n_reviews END AS n_reviews,
    rating.avg_rating
FROM 
    books b
LEFT JOIN (
    SELECT 
        book_id,
        COUNT(review_id) AS n_reviews
    FROM reviews
    GROUP BY book_id
) AS review
    ON review.book_id = b.book_id
LEFT JOIN (
    SELECT 
        book_id,
        AVG(rating) AS avg_rating
    FROM ratings
    GROUP BY book_id
) AS rating
    ON rating.book_id = b.book_id
"""
book_reviews_ratings = pd.read_sql(query, engine)
book_reviews_ratings

book_id  n_reviews  avg_rating
0        652          2    4.500000
1        273          2    4.500000
2         51          5    4.250000
3        951          2    4.000000
4        839          4    4.285714
..       ...        ...         ...
995       64          4    4.230769
996       55          2    5.000000
997      148          3    3.428571
998      790          2    3.500000
999      828          2    3.000000

[1000 rows x 3 columns]

Conclusion: el numero de reviews dependera del libro, pero se puede notar que minimo se tienen 2 reviews por libro, pero no se puede garantizar una tendencia.

## <a id='toc3_4_'></a>[Editorial que ha publicado el mayor número de libros con más de 50 páginas](#toc0_)

In [9]:
query = """
SELECT 
    p.publisher,
    COUNT(b.book_id) AS n_books
FROM 
    books b
LEFT JOIN 
    publishers AS p 
    ON b.publisher_id = p.publisher_id
WHERE 
    b.num_pages > 50
GROUP BY p.publisher
ORDER BY n_books DESC
LIMIT 1;
"""
top_publisher = pd.read_sql(query, engine)
top_publisher

publisher  n_books
0  Penguin Books       42

Conclusion: La editorial Penguin es la que ha publicado la mayor cantidad de libros que contienen mas de 50 paginas, en total ha publicado 42 libros, por lo que es una editorial que se enfoca mas en publicaciones extensas que excluyen folletos o historias cortas. 

## <a id='toc3_5_'></a>[Autor con la Calificación Promedio Más Alta de Sus Libros (Con al Menos 50 Calificaciones)](#toc0_)

In [10]:
query = """
WITH book_ratings AS (
    SELECT 
        b.author_id,
        b.book_id,
        COUNT(r.rating_id) AS num_ratings,
        AVG(r.rating) AS avg_rating
    FROM books AS b
    LEFT JOIN ratings AS r
        ON b.book_id = r.book_id
    GROUP BY b.author_id, b.book_id
    HAVING COUNT(r.rating_id) >= 50
),
author_ratings AS (
    SELECT 
        a.author,
        AVG(br.avg_rating) AS avg_author_rating
    FROM book_ratings AS br
    LEFT JOIN authors AS a 
        ON br.author_id = a.author_id
    GROUP BY a.author
)
SELECT 
    author, 
    avg_author_rating
FROM author_ratings
ORDER BY avg_author_rating DESC
LIMIT 1;
"""

df_top_author = pd.read_sql(query, engine)
df_top_author

author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844

Conclusion: El autor con la calificacion promedio mas alta es J.K Rowling con un rating promedio de 4.28. Esto desmuestra que esta autora tiene una gran recepcion a su trabajo por parte del publico.

## <a id='toc3_6_'></a>[Número Promedio de Reseñas de Texto entre Usuarios que Calificaron Más de 50 Libros](#toc0_)

In [11]:
query = """
WITH user_ratings AS (
    SELECT 
        username,
        COUNT(book_id) AS n_reviews
    FROM ratings AS r
    GROUP BY username
    HAVING COUNT(book_id) > 50
),
user_reviews AS (
    SELECT 
        r.username,
        COUNT(r.review_id) AS n_reviews
    FROM user_ratings AS ur
    LEFT JOIN reviews AS r
        ON ur.username = r.username
    GROUP BY r.username
)
SELECT 
    AVG(n_reviews) AS avg_reviews_per_user
FROM user_reviews AS ur;
"""
avg_reviews_per_user = pd.read_sql(query, engine)

avg_reviews_per_user

avg_reviews_per_user
0             24.333333

Conclusion: Los usuarios que calificaron mas de 50 libros dejan en promedio 24 reseñas de texto, por lo cual se puede notar que tienen una gran participacion. 